eidr/index.go

```diff
@@ -0,0 +1,420 @@
+		for _, field := range []string{
+			"Clip",
+			"Edit",
+			"Season",
+			"EpisodeInfo",
+			"Manifestation",
+		} {
+			e, err := extraobj.GetLink(field)
+			if err != nil {
+				continue
+			}
+			eobj, err := i.store.Get(e.Cid)
+			graph := meta.NewGraph(i.store, eobj)
+
+			switch field {
+			case "EpisodeInfo":
+				episodeobj := episode{}
+				parent, err := graph.Get("Parent", "@value")
+				if err != nil {
+					return err
+				}
+				var parentid doi.ID = doi.ID(parent.(string))
+				baseObject.parentId = &parentid
+
+				// sequenceinfo is an optional field
+				// there may be several items per episode
+				sepcid, err := eobj.GetLink("SequenceInfo")
+				if err == nil {
+					episodeobj.SequenceInfo = &SequenceInfo{}
+					seqobj, err := i.store.Get(sepcid.Cid)
+					if err != nil {
+						return err
+					}
+					var m map[string]interface{}
+					err = seqobj.Decode(&m)
+					if err != nil {
+						return err
+					}
+					for field, _ := range m {
+						if field[0:1] == "@" {
+							continue
+						}
+						sub, err := seqobj.GetLink(field)
+						if err != nil {
+							return err
+						}
+						subobj, err := i.store.Get(sub.Cid)
+						vs, err := subobj.GetString("@value")
+						if err != nil {
+							return err
+						}
+						// domain is optional
+						d, _ := subobj.GetString("domain")
+						seq := &Sequence{
+							Value:  vs,
+							Domain: d,
+						}
+						switch field {
+						case "DistributionNumber":
+							episodeobj.SequenceInfo.DistributionNumber = seq
+							break
+						case "HouseSequence":
+							episodeobj.SequenceInfo.HouseSequence = seq
+							break
+						}
+					}
+				}
+				extratype = episodeobj
+			}
+		}
+	}
+
+	// INSERT INTO INDEX DB
+	// TODO: rollback on unsuccessful reset
+
+	// insert baseobject single item fields
+	_, err = i.db.Exec(
+		"INSERT INTO baseobject (doi_id, structural_type, referent_type, resource_name, resource_name_lang, resource_name_class, status) VALUES ($1, $2, $3, $4, $5, $6, $7)",
+		baseObject.id,
+		baseObject.simple["StructuralType"],
+		baseObject.simple["ReferentType"],
+		baseObject.resourceName,
+		baseObject.resourceNameLang,
+		baseObject.resourceNameClass,
+		baseObject.simple["Status"])
+	if err != nil {
+		return err
+	}
+	// insert multiple value base object fields
+	for _, org := range baseObject.associatedOrg {
+		_, err = i.db.Exec(
+			"INSERT INTO org (id, idtype, display_name, role, base_doi_id) VALUES ($1, $2, $3, $4, $5)",
+			org.ID,
+			org.IDType,
+			org.DisplayName,
+			org.Role,
+			baseObject.id)
+		if err != nil {
+			return err
+		}
+	}
+	for _, altid := range baseObject.alternateID {
+		_, err = i.db.Exec(
+			"INSERT INTO alternateid (id, type, domain, relation, base_doi_id) VALUES ($1, $2, $3, $4, $5)",
+			altid.ID,
+			altid.Type,
+			altid.Domain,
+			altid.Relation,
+			baseObject.id)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	// process extrametadata
+	if extratype != nil {
+		t := reflect.TypeOf(extratype)
```

Comment:
Is it necessary to use reflection here?

Reply:
We switch on the type of the object. What alternative would you suggest?

Reply:
A [type switch](https://golang.org/ref/spec#Type_switches):

```go
switch t := extratype.(type) {
case episode:
	// t has type 'episode' here
}
```

Reply:
that's new to me. Thanks.